src/target/target.cc

```diff
@@ -162,14 +314,164 @@ Target Target::Create(const String& target_str) {
+      ObjectRef key, val;
+      try {
+        key = TargetNode::ParseAttr(kv.first, *info.key);
+      } catch (const tvm::Error& e) {
+        LOG(FATAL) << "Error occurred when parsing a key of the dict: " << kv.first
+                   << ". Details:\n"
+                   << e.what();
+      }
+      try {
+        val = TargetNode::ParseAttr(kv.second, *info.val);
+      } catch (const tvm::Error& e) {
+        LOG(FATAL) << "Error occurred when parsing a value of the dict: " << kv.second
+                   << ". Details:\n"
+                   << e.what();
+      }
+      result[key] = val;
+    }
+    return Map<ObjectRef, ObjectRef>(result);
+  }
+  LOG(FATAL) << "Unsupported type registered: \"" << info.type_key
+             << "\", and the type given is: " << obj->GetTypeKey();
+  throw;
+}
+
+Target Target::FromConfig(const Map<String, ObjectRef>& config_dict) {
+  const String kKind = "kind";
+  const String kTag = "tag";
+  const String kKeys = "keys";
+  std::unordered_map<std::string, ObjectRef> config(config_dict.begin(), config_dict.end());
+  ObjectPtr<TargetNode> target = make_object<TargetNode>();
+  // parse 'kind'
+  if (config.count(kKind)) {
+    const auto* kind = config[kKind].as<StringObj>();
+    CHECK(kind != nullptr) << "AttributeError: Expect type of field 'kind' is string, but get: "
+                           << config[kKind]->GetTypeKey();
+    target->kind = TargetKind::Get(GetRef<String>(kind));
+    config.erase(kKind);
+  } else {
+    LOG(FATAL) << "AttributeError: Field 'kind' is not found";
+  }
+  // parse "tag"
+  if (config.count(kTag)) {
+    const auto* tag = config[kTag].as<StringObj>();
+    CHECK(tag != nullptr) << "AttributeError: Expect type of field 'tag' is string, but get: "
+                          << config[kTag]->GetTypeKey();
+    target->tag = GetRef<String>(tag);
+    config.erase(kTag);
+  } else {
+    target->tag = "";
+  }
+  // parse "keys"
+  if (config.count(kKeys)) {
+    std::vector<String> keys;
+    // user provided keys
+    const auto* cfg_keys = config[kKeys].as<ArrayNode>();
+    CHECK(cfg_keys != nullptr)
+        << "AttributeError: Expect type of field 'keys' is an Array, but get: "
+        << config[kTag]->GetTypeKey();
+    for (const ObjectRef& e : *cfg_keys) {
+      const auto* key = e.as<StringObj>();
+      CHECK(key != nullptr) << "AttributeError: Expect 'keys' to be an array of strings, but it "
+                               "contains an element of type: "
+                            << e->GetTypeKey();
+      keys.push_back(GetRef<String>(key));
+    }
+    // add device name
+    if (config_dict.count("device") && config_dict.at("device")->IsInstance<StringObj>()) {
+      keys.push_back(Downcast<String>(config_dict.at("device")));
+    }
+    // add default keys
+    for (const auto& key : target->kind->default_keys) {
+      keys.push_back(key);
+    }
+    // de-duplicate keys
+    target->keys = DeduplicateKeys(keys);
```

Comment:
I'm not sure if this part could catch exceptions. For example, what's the expected targets from the following configs?
```
{"kind": "llvm", "keys": ["cpu", "arm_cpu"], "device": "arm_cpu"}
{"kind": "llvm", "keys": ["gpu"]}
```
Also, if the config is
```
{"kind": "llvm", "keys": ["arm_cpu", "cpu"]}
```
Should `"device": "arm_cpu"`  be added to the target?

To me those cases should be covered in the unit test as well.

Reply:
The original design of "-keys" semantic is that:

1) user provided keys comes first
2) add `device_name` if given
3) add default keys in the `target_kind`
4) deduplicate the keys, only preserve the first appearance of each key.

Therefore, the expected keys of the target created is `["cpu", "arm_cpu"]`.

Reply:
Then are the first two examples valid? While I'm not 100% for sure about the first one, I think the second one is definitely invalid, as its keys will be `["gpu", "cpu"]`.

Also, what happen if `device_name` is empty as the third example?

Reply:
**In response to the second example.** We can see that the second example exposes an extreme case that the keys can be semantically unnatural, i.e. typically there are should not be targets on both cpu and gpu. It is typically impossible.

To keep the sanity of the workflow, a check must be conducted somewhere; or we have to assume the users are all "grown adults" who can write correct target (which IMO is still valid assumption). IIUC the problem is that should target creation do such check? IMHO, however, we don't need it to be constant magic strings fixed in C++ side of "target", for the following reasons:
* Imagine an external developer who installs TVM using pip/conda, and he just wanted to write all operators in python using his own specific keys for his specific arm device. Are we harsh enough to ask him to recompile the entire TVM codebase, just to allow this key to exist? It doesn't really let you gain anything, but hurts flexibility.
* It is hardly possible to enumerate all invalid cases. As those two examples you proposed, they are "somewhat invalid" in different ways, reversed key order, semantically incompatible keys, etc. Have they really covered all the cases? IIUC, no, not really.  I could imagine thousands of possible combinations that are semantically wrong. For example, "mtriple", "march", "mcpu" are incompatible. It is just impossible to enumerate all of them.

**Syntactical vs semantical.** We aim at solving syntactical problems in this series of PRs, but as I have mentioned in this post, semantical issues are never possible to be enumerated or solved, and shouldn't be solved under the scope of target. What we really want is a high-level wrap of target creation that makes semantic checking possible, or the "target tag" system that guarantees correct outcome.

Thank you for the very constructive comments! Those comments really made me rethink and understand more deeply into all those problems. Thank you Cody!

Reply:
Fair enough. Then I would just have 3 points accordingly:
- We don't aim to have a comprehensive semantic checking in the target parser. Instead, we only insist the basic semantic for users that do not want to spend time on investigating how target should be set. For example, `{"kind": "llvm"}` for X86, `{"kind": "llvm", "device": "arm_cpu"}` for ARM.
- Assuming the error handling that the community is working on would be mature shortly, the target with incorrect semantic should be caught somewhere in the build pipeline and a proper error message will be delivered to users.
- As I mentioned before in the online meetup, we really need a document to teach everyone how to set up a proper target to fit the environment. We should open an issue to track the progress of this series of target updating and documentation sould be the last step.

Reply:
> {"kind": "llvm"} for X86, {"kind": "llvm", "device": "arm_cpu"} for ARM.

This can be done via the helper functions we already have.

> the target with incorrect semantic should be caught somewhere in the build pipeline and a proper error message will be delivered to users.

I agree with the spirit, and I did see some mysterious error info when target is incorrect. This is the direction our community should work towards :-)

> we really need a document to teach everyone how to set up a proper target to fit the environment

I strongly agree with it. I could see meaning of those target is rarely documented and somehow becomes mysterious. There are cases of misuse in the current codebase too.